linera-core/src/chain_worker/state/mod.rs

```diff
@@ -305,12 +305,12 @@ where
     }
 
     /// Returns an error if the block requires a blob we don't have.
-    /// Looks for the blob in: chain manager's pending blobs and storage.
+    /// Looks for the blob in: chain manager's proposed blobs and storage.
```

Comment:
IIUC in storage we have blobs from _confirmed_ blocks and in the `proposed_blobs` we have blobs that haven't been validated even (they'd be part of `locked_blobs`). Isn't it possible that the blobs from `proposed_blobs` will be forgotten (if we confirm another proposal)? Yet the caller of this method might be told that we do have all the blobs (which we'll later delete). Isn't that a problem?

Reply:
> blobs that haven't been validated even

(They have been validated by _us_, but not yet by a quorum, yes.)

I'm also a bit confused, yes.

Whenever we set the proposed block, we need to make sure that the blobs we store with the proposal are exactly the ones published by it. (Whether they were provided by the client or we already had them. But maybe we should just require for now that the client provide them all regardless.)

And analogously for the locked block: We need to store with it all the blobs that it publishes _or_ reads! (Read blobs that we already have in storage we can probably omit, since we currently never remove blobs from storage, and since that might end up being too many blobs for the chain manager.)

Reply:
(Ah, I guess in the second case we never need to store any blobs in the chain manager that aren't being published by this block, because we presumably already asked for and processed a confirmed block that used them?)

Reply:
I guess now that we don't have `proposed_blobs` anymore and we just point to the blobs in the `BlockProposal`, we should be fine here, right? Or am I missing something?

Reply:
I think you're right:
* We should check that the blobs provided in the proposal are all published by that block (and don't contain duplicates).
* The published blobs that aren't provided in the proposal must be confirmed for us locally.

That's both already the case, is it?

Reply:
I believe so. We create the proposal's blobs based on the blobs being _published_ by that block. We also enforce that all blobs being published must be provided in the proposal, which means they must be in some _pending_ location locally when proposing the block.
Right now, if we're trying to publish a blob again that is already confirmed in storage, we do technically need to add that blob to pending. Later on that'll be optimized away as we will ignore blobs that are already in storage, but maybe we can do that optimization earlier as well. Let me try that.

Reply:
Right, of course, my bad. I'll add the duplicated blobs check!

Reply:
Actually, on the server side we also always get the blobs for the block either by calling [published_blob_ids](https://github.com/linera-io/linera-protocol/blob/main/linera-chain/src/data_types.rs#L68) or [required_blob_ids](https://github.com/linera-io/linera-protocol/blob/main/linera-chain/src/data_types.rs#L950). So I don't think we'll have duplicates.
The question is: should these functions error out if they detect a duplicate? For `published_blob_ids` I would say maybe, but for `required_blob_ids` I don't think so

Reply:
Is it done already? if so, can you point to the change that does what's discussed?